orttraining/orttraining/python/training/optim/lr_scheduler.py

```diff
@@ -0,0 +1,290 @@
+import math
+
+from . import config
+
+
+class _LRScheduler(object):
+    r"""Base class for implementing custom learning rate schedulers
+
+    Schedulers can be either stateful or stateless.
+    Stateless implementations can only rely on information available at
+    :py:class:`.TrainStepInfo` and :py:attr:`.optimizer_config`.
+    Stateful, on the other hand, can store more parameters by overriding
+    the constructor.
+
+    In both cases, once the scheduler is configured, no user code is needed
+    to update learning rate during each train step.
+
+    NOTE: This class should never be instantiated, but used as an abstract class.
+
+    NOTE: Current implementation doesn't support 'lr' within :py:attr:`param_groups` entries.
+
+    Args:
+        optimizer_config (optim._OptimizerConfig): optimizer config.
+            One of :py:class:`.optim.Adam`, :py:class:`.optim.Lamb` or :py:class:`.optim.SGD`.
+    """
+
+    def __init__(self, optimizer_config):
+        assert isinstance(optimizer_config, config._OptimizerConfig),\
+            "optimizer_config must be :py:class:`.optim._OptimizerConfig"
+        for group in optimizer_config.param_groups:
+            assert 'lr' not in group, "'lr' is not supported inside param_groups"
+        self.optimizer_config = optimizer_config
+        self._last_lr = []
+
+    def get_lr(self, train_step_info):
+        r"""Returns a list of learning rate
+
+        Args:
+            train_step_info (:py:class:`.TrainStepInfo`): runtime info for current training step
+
+        Returns:
+            ordered :py:obj:`list` of learning rates.
+                The first entry is the default learning rate and
+                    the remaining, refer to each :py:class:`optim._OptimizerConfig`s parameter group.
+            NOTE: Currently, only the default learning rate is supported,
+                which implies returning a list with a single value.
+        """
+        raise NotImplementedError
+
+    def get_last_lr(self):
+        r""" Return last computed learning rate by LR Scheduler"""
+        return self._last_lr
+
+    def _step(self, train_step_info):
+        r"""Private method called to update :py:class:`optim._OptimizerConfig` learning rate
+
+        NOTE: This class should never be called by the user.
+        """
+        new_lr = self.get_lr(train_step_info)
+        self._last_lr = new_lr
+        self.optimizer_config.lr = new_lr[0]
+
+
+class ConstantWarmupLRScheduler(_LRScheduler):
```

Comment:
For the following LRSchedulers, is there a way to tell if they are stateful or stateless?

Reply:
The class description and/or implementation are the only official way
Looking the LRSCheduler constructor we can make an educated guess. A stateless class doesn't need to receive any input in the constructor, as it shouldnt store state from external sources. However, that doesnt mean the LRScheduler is not initializing state internally without external information.